### PR TITLE
feat(DataTable): make size and columns customizable via toolbar menu

### LINKED ISSUE #4687: Table settings menu
### Summary

We are requesting a data table settings menu component based on the design mocked up here https://www.carbondesignsystem.com/components/data-table/usage. 

Screen Shot 2019-11-14 at 10 37 25 AM

In addition to displaying row height it would also allow the user to hide and show columns. 

**Proposed design:**

image

image

image

image

### Justification

Over a third of IBM Cloud tables are based on a Softlayer table that offer edit column functionality. If we don't add this functionality to the Carbon data table, users will lose that existing functionality when our platform moves to Carbon X. We also have research showing that users would like to be able to edit their columns, especially on heavily used tables such as device or resource list (we can share quotes as needed). For example, one user had an internal site that only used private IPs, and he wanted to hide public IPs. Editable columns were also a frequently mentioned functionality by platform and service teams alike.

### Desired UX and success metrics

The desired UX is that users can see the columns that make sense for the job(s) they are trying to complete. For users that already enjoy this feature on Softlayer tables, the goal is to preserve existing functionality. We will track clicks on the setting menu as well as the number of users who edit columns on a given table.

### "Must have" functionality

- A user can select row height options
- A user can select and deselect columns
- User can scroll if there are more column options than fit the dropdown
- Must meet accessibility standards
- A designer can choose to ONLY include row height or edit columns 
- After settings menu reaches it's **max height** (496px), edit columns section should **scroll**
- Responsive (see design above)

Nice to have: 
- **_After_** a user closes the settings menu, any checked columns move to top of edit columns section and appear there the next time they open the menu.
- User can reset to default (design needs to add this to mockup)

### Specific timeline issues / requests

This work is related to the CloudPAL-sponsored project to contribute back a more robust data table to Carbon by mid-December, as part of Carbon X / CloudPAL adoption. As such, we have a very tight timeline. We would like teams to be able to start consuming these tables in January.

Our team is working closely with Carbon design and development and trying to anticipate accessibility concerns.

### Available extra resources

Holman from ICED development squad in Toronto, possibly CloudPAL developers. Holman is currently working on developing this feature.

### PR DESCRIPTION
**Pull request by holmansze**
Closes #4687

Add a settings menu to the toolbar to enable customization of table size (row height) and columns (set of rendered columns) by end user.

#### Changelog

**New**
- add `TableToolbarSettingSize` and `TableToolbarSettingColumns` for the settings menu

**Changed**
- update `TableToolbarMenu` to support settings menu
- update `DataTable` to honor settings via the menu

**Removed**
- nil

#### Testing / Reviewing
- in the storybook, go to DataTable > with toolbar (settings), click the settings icon (the gear icon between the search and primary button) to customize table size and columns
- all new/changed code have 100% unit tests coverage

### CODE REVIEW

**Review by asudoh — COMMENTED** [review]
Thanks @holmansze for jumping in! Just one thing - Great if we can avoid adding features to the state manager (`<DataTable>`). Seems that we can have those settings managed in application's state, right?

**Comment by holmansze** [comment]
@asudoh - I think if we don't use internal state for size and columns, the storybook won't work and the component will look broken.  Let's say you specify `size=short` and `selectedColumns=['columnA', 'columnB']`, the table is initially rendered as short with 2 columns.  Then you open the settings menu to change size and columns, the table won't update.

To my understanding, other features (row selection, expansion, search, sort) are all implemented with this approach (use prop to populate initial state).  Correct?

**Comment by asudoh** [comment]
Thank you for sharing your thoughts @holmansze! The issue you pointed out can typically be solved by using story-only components (e.g. [here](https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/react/src/components/FileUploader/stories/drop-container.js) used from [here](https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/react/src/components/FileUploader/FileUploader-story.js#L177-L188)).

File: packages/react/src/components/DataTable/TableToolbarMenu.js [hunk 1]
```diff
@@ -21,6 +22,41 @@ const TableToolbarMenu = ({
...
 }) => {
+  const [focused, setFocus] = useState(0);
+
+  useEffect(() => {
+    updateMenuFocus();
+  });
+
+  const updateMenuFocus = () => {
+    const nodes = [
+      ...document.querySelectorAll('[data-table-toolbar-focusable]'),
```
Comment (asudoh) [inline packages/react/src/components/DataTable/TableToolbarMenu.js:33]: Any way to limit the scope of DOM search?

File: packages/react/src/components/DataTable/TableToolbarMenu.js [hunk 2]
```diff
@@ -21,6 +22,41 @@ const TableToolbarMenu = ({
...
+    ];
+    if (nodes && focused >= 0 && nodes[focused]) {
+      nodes[focused].focus();
+    }
+  };
+
+  const handleMenuItemFocus = evt => {
+    const nodes = [
+      ...document.querySelectorAll('[data-table-toolbar-focusable]'),
```
Comment (asudoh) [inline packages/react/src/components/DataTable/TableToolbarMenu.js:42]: Any way to limit the scope of DOM search?

**Review by asudoh — COMMENTED** [review]
On a second thought, your current approach with state management is fine to me. Basically my concern was it may make the review progress longer.

**Comment by holmansze** [comment]
Thanks @asudoh 
I guess it is not just a storybook problem, but the usability of the component in general.

I think if only props are used to manage the settings, we are assuming the consumer always manages/persists the table settings.

For instance, I may implement a page that uses `DataTable`, and my application doesn't persist table settings state.  I always default to size `normal` with all columns shown.  When user makes changes via the settings menu I expect the component to update, though I won't be persisting the settings.

(Oh I saw your comment just now, please disregard this one.)

**Comment by joshblack** [comment]
Hey @holmansze! 👋 Thanks so much for taking the time to put this Pull Request together. Like we were discussing in the data table channel, we do need design to sign off on the issue linked before moving forward with reviews. Would it make sense to hold off until we get that design alignment?

We could also investigate how to layer this on top of the existing Data Table instead of baking it into the component. This might be a smoother path forward and lead to less churn. Happy to chat about this too if you're interested 👍

**Comment by holmansze** [comment]
@joshblack - once I get your green light I will address Akira's comments and resolve the conflicts.  Let me know.  Thanks.

**Comment by joshblack** [comment]
Hey @holmansze! 👋 Going to close this out since I think you're PR'ing this stuff into Cloud PAL? Let me know if not and I'll be happy to re-open!

**Pull request closed without merging**